src/Common/Exception.cpp

```diff
@@ -280,7 +280,7 @@ bool Exception::isErrorCodeImportant() const
 Exception::~Exception()
 try
 {
-    if (logged != nullptr && !logged->load(std::memory_order_relaxed) && isErrorCodeImportant())
+    if (logged != nullptr && !logged->load(std::memory_order_relaxed) && isErrorCodeImportant() && isLoggingEnabled())
```

Comment:
This is what was preventing the "race condition" in https://github.com/ClickHouse/ClickHouse/issues/82059 due to different static object destruction order from appearing before. Needs to be added back in the revert-revert PR or add something similar